python/rainbow/simulators/proximal_contact/gauss_seidel_solver.py

```diff
@@ -0,0 +1,124 @@
+            )  # Used to keep the old values and compute the change in values
+            r_b = self.r[block]
+            b_b = self.b[block]
+
+            # By definition
+            #       z = x - r (J WJ^T x  + b)
+            #         = x - r ( A x  + b)
+            # We use
+            #        w =  WJ^T x
+            # so
+            #       z  = x - r ( J w  + b)
+            z_b = x_b - np.multiply(r_b, (self.J.dot(w)[block] + b_b))
+
+            # Solve:         x_n = prox_{R^+}( x_n - r (A x_n + b) )
+            x_b[0] = np.max([0.0, z_b[0]])
+
+            # Solve:         x_f = prox_C( x_f - r (A x_f + b))
+            x_b[1], x_b[2], x_b[3] = self.friction_solver(z_b[1], z_b[2], z_b[3], mu_k, x_b[0])
+            # Put updated contact forces back into solution vector
+            self.x[block] = x_b
+            # Get the change in the x_block
+            np.subtract(x_b, delta, delta)
+            # Update w
+            w += self.WJT.tocsr()[:, block].dot(delta)
+
+
+@nb.njit(parallel=True, nogil=True, cache=True)
+def sweep_worker(color_group, J, WJT, b, mu, r, x, w, delta_ws, friction_solver):
+    """ The worker function of the parallel gauss seidel algorithm.
+
+    Args:
+        color_group (ArrayLike): The color group, the value is  the block location, containing the start and end index.
+        J (ArrayLike): The contact jacobi matrix.
+        WJT (ArrayLike): The WJ^T matrix, here W = M^{-1}, and the M is the mass matrix.
+        b (ArrayLike): b = Ju^t + ∆t JM^{-1}h + EJu^t
+        mu (float): The coefficient of friction.
+        r (float): r-factor value.
+        x (ArrayLike): The current contact force.
+        w (ArrayLike): The WJT.dot(x).
+        delta_ws (ArrayLike): A array of the delta_w, each delta_w = WJT.dot(delta_x), here delta_x is the change of the contact force.
+        friction_solver (callable): The proximal operator of friction cone function.
+
+    Returns:
+        (ArrayLike, ArrayLike): The new contact force and the new delta_ws.
+    """
+    for i in nb.prange(len(color_group)):
+        block_start, block_end = color_group[i]
+        block = np.arange(block_start, block_end)
+        x_b = x[block]
+        r_b = r[block]
+        b_b = b[block]
+
+        delta = x_b.copy()
+        z_b = x_b - np.multiply(r_b, (J.dot(w)[block] + b_b))
+
+        x_b[0] = np.max(np.array([0.0, z_b[0]]))
+
+        mu_k = mu[block_start // 4]
+        x_b[1], x_b[2], x_b[3] = friction_solver(z_b[1], z_b[2], z_b[3], mu_k, x_b[0])
+
+        np.subtract(x_b, delta, delta)
+        delta_w = WJT[:, block].dot(delta)
+
+        x[block] = x_b
+        delta_ws[block_start//4] = delta_w
+
+    return x, delta_ws
+
+
+class ParallelGaussSeidelSolver(SolverInterface):
+    """ Parallel Gauss Seidel Solver.
+    """
+    def __init__(self, J, WJT, b, mu, friction_solver, engine, stats, debug_on, prefix):
+        super().__init__(J, WJT, b, mu, friction_solver, engine, stats, debug_on, prefix)
```

Comment:
Should inheritance not take core of this?

Reply:
I will remove it.